### ISSUE TITLE
Boolean fields in a mutation's input type are dropped when false

### ISSUE THREAD

**Issue by redjohn** [absinthe-graphql/absinthe_relay#35]
I'm working on a project where we want to supply a boolean value as a field on a mutation's input type, for example:

```elixir
    payload field :create_user do
      input do
        field :name, non_null(:string)
        field :admin, non_null(:boolean)
      end
      output do
        field :user, :user
      end
      resolve &Resolver.User.create/2
    end
```

With the Absinthe v1.2 branch (and the corresponding branch for Absinthe Relay), if the supplied value for `admin` is `false`, it gets dropped from the `args` before it gets to the resolver. I've set up an example Phoenix app to illustrat the issue. This is what is logged:

```
[info] POST /docs
[info] Variables: %{"input" => %{"admin" => false, "clientMutationId" => "0", "name" => "Not Test User"}}
[debug] GraphQL Document:
mutation testing($input:CreateUserInput!) {
  createUser(input:$input) {
    user {
      id
      name
      admin
    }
  }
}

[info] args for create: %{client_mutation_id: "0", name: "Not Test User"}
[info] Sent 200 in 24ms
```

When I define a mutation without using any of the Relay Notation, the `admin` value is not dropped.

Mutation definition:
```elixir
    field :other_create_user, type: :user do
      arg :name, non_null(:string)
      arg :admin, non_null(:boolean)
      resolve fn args, ctx ->
        with {:ok, user} <- Resolver.User.create(args, ctx),
          do: {:ok, user[:user]}
      end
    end
```

Log:
```
[info] POST /docs
[info] Variables: %{"admin" => false, "name" => "Not a Test User"}
[debug] GraphQL Document:
mutation testing($admin: Boolean!, $name:String!) {
  otherCreateUser(admin: $admin, name:$name) {
    id
    name
    admin
  }
}

[info] args for create: %{admin: false, name: "Not a Test User"}
[info] Sent 200 in 167ms
```

As a result, I think the issue must be somewhere inside the Absinthe Relay support. All cases work with Absinthe 1.1 and the corresponding version of Absinthe Relay. You can see my full app at https://github.com/redjohn/absinthe_relay_boolean_mutation_field. There are branches named 1.1 and 1.2 with tests that illustrate the issue.

I'm willing to help fix it, but could use a pointer about where to start looking in the code. Hope this is helpful.

**Comment by bruce** [comment]
Thanks @redjohn -- we'll look into this. My guess (without looking) is it's due to some change in the way we wrap resolvers for Relay (that is, that your intuition is correct).

Pinging @benwilson512 on this as well, since this effectively blocks release of v1.2 for Relay support.

**Comment by benwilson512** [comment]
You're the best person on earth for making such thorough test cases! I'm investigating now :)

**Comment by benwilson512** [comment]
Found the issue. It wasn't relay specific, but it was specific to input objects, which was used in the relay case and not the other cases. Basically, I was trying to filter out `nil` inputs by doing

```elixir
|> Enum.filter(fn {_, v} -> v end)
```

Whoops.

Now it's
```
|> Enum.reject(&match?({_, nil}, &1))
```

Will have a fix pushed shortly, thanks!

**Comment by benwilson512** [comment]
This is fixed on the v1.2 branch of absinthe. It'll be included in the v1.2 release later today.

Thank you!

**Comment by redjohn** [comment]
Thanks for the quick response guys :)
